apps/system/lockscreen/js/lockscreen_inputpad.js

```diff
@@ -111,25 +116,111 @@
       case 'lockscreen-inputappclosed':
         this.updatePassCodeUI();
         break;
+      case 'touchstart':
+        // TODO: ignore while on error timeout
+        touch = evt.changedTouches[0];
+        target = document.elementFromPoint(touch.clientX, touch.clientY);
+        key = this._keyForTarget(target);
+        if (!key) {
+          break;
+        }
+        if (this.lastTouched) {
+          this._makeKeyInactive(this.lastTouched);
+        }
+        this._makeKeyActive(target);
+        this.lastTouched = target;
+        if (this.states.padVibrationEnabled &&
+          !this.states.passCodeErrorTimeoutPending) {
+          navigator.vibrate(this.configs.padVibrationDuration);
+        }
+        break;
+      case 'touchmove':
+        // TODO: ignore while on error timeout
+        touch = evt.changedTouches[0];
+        target = document.elementFromPoint(touch.clientX, touch.clientY);
+        key = this._keyForTarget(target);
+        if (target.textContent !== this.lastTouched.textContent) {
+          if (this.lastTouched) {
+            this._makeKeyInactive(this.lastTouched);
+          }
+          if (key) {
+            this._makeKeyActive(target);
+            this.lastTouched = target;
+          } else {
+            this.lastTouched = null;
+          }
+        }
+        break;
+      case 'touchend':
       case 'click':
-        var key = evt.target.dataset.key;
-        if (!key &&
-            ('div' === evt.target.tagName.toLowerCase() &&
-             'a' === evt.target.parentNode.tagName.toLowerCase())
-           ) {
-          key = evt.target.parentNode.dataset.key;
+        // TODO: ignore while on error timeout
+        if (evt.type == 'touchend') {
+          evt.preventDefault();  // prevent the 'click'
+          touch = evt.changedTouches[0];
+          target = document.elementFromPoint(touch.clientX, touch.clientY);
+        } else {  // 'click'
+          target = evt.target;
+        }
+        key = this._keyForTarget(target);
+        if (this.lastTouched) {
```

Comment:
Added comments throughout the code. Refactored the variable to .lastTouchedKey.

In short: it's used to determine whether the finger has moved to a new key since the last event. If and only if that is the case, lastTouchedKey must be made inactive, and the new key must be made active.